### DISCUSSION TITLE
Kitty remote control and kitten wrapper

### DISCUSSION THREAD

**Discussion by doctorfree** [kovidgoyal/kitty#7104] [Show and tell]
I've written a `kitty` remote control and `kitten` wrapper that enables quick and easy scripted control of some aspects of `kitty`.

See the [kitty-control readme](https://github.com/doctorfree/kitty-control#readme) for an overview, installation instructions, examples, and usage. At first I just wanted an easy way to switch `kitty` configuration when I moved from a small screen (laptop) to a large screen (tv). To do this I wrote a `kitten` that loads a specified `kitty` configuration without needing to quit `kitty` and restart with another config.

Then things escalated quickly and I added several other quick and easy ways to manipulate `kitty` using both remote control commands and kittens. I keep adding more stuff. Most of what `kitty-control` does can be done with `kitty` directly if you know the right command. It acts mainly as a convenience but also adds some functionality. In addition to the `kitty` control features, `kitty-control` detects if either `kitty` or `kitty-control` have available updates and presents an option to perform the upgrade.

Included with the `kitty-control` installation are `kitty-control` itself, the `read_config.py` kitten, and a UNIX style man page.

Typical usage is from the command line but when invoked without arguments an interactive menu is presented. The main `kitty-control` interactive menu provides dynamically generated menu entries. For example, if either `kitty` or `kitty-control` have an
available upgrade then menu entries to upgrade will be added.

The `kitty-control` main menu:

![Main Menu](https://raw.githubusercontent.com/wiki/doctorfree/kitty-control/control-menu.png)

Check it out and let me know what you think and how it could be improved.

**Comment by kovidgoyal** [comment]
You no longer need the kitten part with: https://github.com/kovidgoyal/kitty/commit/065b17ddbd426fbb430579ab2a5b4de466becff4

**Reply by doctorfree** [reply]
@kovidgoyal  That's great! Thanks.

I will replace my kitten with calls to `load-config-file`. In which version of `kitty` was this feature introduced?

**Reply by kovidgoyal** [reply]
It's in master

**Reply by doctorfree** [reply]
Ok, here is what I'm currently doing in `kitty-control` to check whether I can use the `load-config-file` Kitty command:

```bash
# Check for new load-config-file Kitty command
USE_LOAD_CONFIG=1
if [ "${KITTY_WINDOW_ID}" ]; then
  # Running in a Kitty window
  kitty @ ${SOCKET} load-config-file 2>/dev/null
  [ $? -eq 0 ] || USE_LOAD_CONFIG=
else
  # Not sure how to check without a Kitty window so use read_config.py
  USE_LOAD_CONFIG=
fi
```
Perhaps there is a simpler approach, this is the first pass.

Once `load-config-file` is integrated into a release I can check for the release and upgrade if necessary.

I see where there has been some discussion of how to detect if running in a Kitty window. I'll read up on that.

Anyway, this first pass at using `load-config-file` seems to be working. Thanks!

**Reply by doctorfree** [reply]
Never mind, I'm still working on this. I think it might be better to do the check and invocation in Python so I may rework this.

**Comment by doctorfree** [comment]
[Version 1.0.9](https://github.com/doctorfree/kitty-control/releases/latest) of `kitty-control` now uses the `load-config` built-in kitten if it is available. In addition, several new features have been added including command line and menu support for opening `kitty` documentation in a browser window using the `action` kitten. Features now include:

- Easily switch a running `kitty` instance configuration without restart
- Set various `kitty` configuration options on the fly including font size, background, foreground, window size, and more
- Diff files or folders, selections are performed using the `ranger` file manager and `fzf` fuzzy finder
- Quickly and easily toggle fullscreen display of the `kitty` terminal window
- Set the `kitty` theme on the fly using the `kitty` **themes** `kitten`
- List colors, tabs, and windows
- Open an interactive menu system to easily browse `kitty-control` features and capabilities
- Display system and terminal information
- Set the level of opacity/transparency on the fly
- Input Unicode characters using the `kitty` **Unicode input** `kitten`
- Select and view an image using the `kitty` **icat** `kitten`
- Select and set a background image using `ranger` and `fzf`
- Set the `kitty` window layout using the `kitty` **goto-layout** `kitten`
- Browse the `kitty` documentation using the `kitty` **action** `kitten`
